### PR TITLE
Add support for key backups using secret storage

### PR DESCRIPTION
**Pull request by jryans**
This add various helpers for creating and restoring key backups where the private key is stored in secret storage.

Part of https://github.com/vector-im/riot-web/issues/11209
Used by https://github.com/matrix-org/matrix-react-sdk/pull/3720

### CODE REVIEW

File: src/client.js [hunk 1]
```diff
@@ -1450,32 +1452,36 @@ MatrixClient.prototype.prepareKeyBackupVersion = async function(
...
-        },
+        auth_data: authData,
         recovery_key: encodedPrivateKey,
     };
 };
 
+/**
+ * Check whether the key backup private key is stored in secret storage.
+ */
+MatrixClient.prototype.isKeyBackupKeyStored = async function() {
```
Comment (turt2live) [inline src/client.js]: it shouldn't need to be async if we don't await

File: src/client.js [hunk 2]
```diff
@@ -1450,32 +1452,36 @@ MatrixClient.prototype.prepareKeyBackupVersion = async function(
...
-            private_key_salt: keyInfo.passphrase.salt,
-            private_key_iterations: keyInfo.passphrase.iterations,
-        },
+        auth_data: authData,
         recovery_key: encodedPrivateKey,
     };
 };
 
+/**
+ * Check whether the key backup private key is stored in secret storage.
```
Comment (turt2live) [inline src/client.js]: For the jsdoc it would be good to have the return type specified explicitly.

File: spec/unit/crypto/backup.spec.js [hunk 3]
```diff
@@ -361,7 +361,7 @@ describe("MegolmBackup", function() {
                     resolve();
                     return Promise.resolve({});
                 };
-                await client.createKeyBackupVersion({
+                client.createKeyBackupVersion({
```
Comment (turt2live) [inline spec/unit/crypto/backup.spec.js]: I'm impressed the tests still pass even though the function is in fact async

File: src/client.js [hunk 4]
```diff
@@ -1624,6 +1634,15 @@ MatrixClient.prototype.restoreKeyBackupWithPassword = async function(
     );
 };
 
+MatrixClient.prototype.restoreKeyBackupWithSecretStorage = async function(
```
Comment (turt2live) [inline src/client.js]: Docs please

File: src/client.js (hunk 1 above)
Comment (jryans) [reply src/client.js]: The inner function is async and / or returns a promise, but `return await` is considered poor form by some as it adds some unnecessary overhead... I thought it best to at least leave the async to suggest the return type of a promise, but also haven't seen a clear style choice for this in code base.

Comment (turt2live) [reply src/client.js]: I roughly read that as "AHHHHHHHHHHHHHHHHHHHHHHHHH" in my head and am fine with the change as-is.

File: spec/unit/crypto/backup.spec.js (hunk 3 above)
Comment (jryans) [reply spec/unit/crypto/backup.spec.js]: This bit triggers the handler installed above which is wrapped in a promise... It's a bit of a monster. 😅

File: src/client.js (hunk 4 above)
Comment (jryans) [reply src/client.js]: Okay, added!

File: src/client.js (hunk 2 above)
Comment (jryans) [reply src/client.js]: Okay, added!

**Pull request merged**